apisix/plugins/zipkin/reporter.lua

```diff
@@ -105,38 +104,68 @@ function _M.report(self, span)
         annotations = span.logs
     }
 
-    local i = self.pending_spans_n + 1
-    self.pending_spans[i] = zipkin_span
-    self.pending_spans_n = i
-end
-
-function _M.flush(self)
-    if self.pending_spans_n == 0 then
-
-        return true
+    self.pending_spans_n = self.pending_spans_n + 1
+    if self.processor then
+        self.processor:push(zipkin_span)
     end
+end
 
-    local pending_spans = cjson.encode(self.pending_spans)
-    self.pending_spans = {}
-    self.pending_spans_n = 0
-
+local function send_span(pending_spans, report)
     local httpc = resty_http.new()
-    local res, err = httpc:request_uri(self.endpoint, {
+    local res = httpc:request_uri(report.endpoint, {
         method = "POST",
         headers = {
             ["content-type"] = "application/json",
         },
         body = pending_spans,
+        keepalive = 5000,
+        keepalive_pool = 5
     })
 
-    -- TODO: on failure, retry?
     if not res then
-        return nil, "failed to request: " .. err
+        return nil, "failed: " .. report.endpoint
     elseif res.status < 200 or res.status >= 300 then
         return nil, "failed: " .. res.status .. " " .. res.reason
     end
 
-    return true
+   return true
+end
+
+function _M.init_processor(self)
+    local process_conf = {
+        name = "zipkin_report",
+        retry_delay = 1,
+        batch_max_size = 1000,
+        max_retry_count = 0,
+        buffer_duration = 60,
+        inactive_timeout = 5
+    }
+
+    local flush = function (entries, batch_max_size)
+        if not entries then
+            return true
+        end
+
+        local pending_spans, err
+        if batch_max_size == 1 then
+            pending_spans, err = cjson.encode(entries[1])
```

Comment:
I am not sure  this code  "cjson.encode_number_precision(16)" can affect encoding.

Reply:
I think it maybe the trace Id is to long.

Reply:
Maybe we can put `cjson.encode_number_precision(16)` into `core.json`? CC @moonming

Reply:
agreed

Reply:
@gy09535 
Therefore we need to use `cjson.safe` instead. Otherwise an error will be thrown when encoding failed.

Reply:
@moonming @gy09535 
I suddenly realized one thing: since `core.json` use `cjson` inside, when we set `encode_number_precision(16)`, we already do the same thing for `cjson`.

Reply:
I think it doesn't matter in this case, it is a batch job and  can not be has error, because for input spans must be serialized.

Reply:
@gy09535 
I don't think so. A proper error handling is better than an unexpected error.

Reply:
@spacewander  After check I find use "cjson" here  can not be cause exception, for encode  just "function"
"ightuserdata(not null value)" "thread" can cause exception, anyway this entity is table, use cjson.safe equals use cjson. I think use cjson is ok.

Reply:
@gy09535 
In similar code of batch processor, we always check the error.
It is meaningless to check the error, if you insist on using  `cjson` directly, since the `encode` method from this module never returns `nil, err`.

Reply:
And, `this entity must be table` doesn't mean ` can not be cause exception`. You can see my example above:
```lua
local cjson = require("cjson")
local x = cjson.new()
local t = {}
t[1] = 1
t[1000] = 1
print(x.encode({a = t}))
# ERROR: /tmp/a.lua:7: Cannot serialise table: excessively sparse array
```

Please! Please! Please! Be careful and never say something bad won't happen. I am tired of telling the same thing too many times.

Reply:
Got it, fixed.